Source/Controllers/Other/SecureBookmarkManager.swift

```diff
@@ -0,0 +1,336 @@
+//
+//  SecureBookmarkManager.swift
+//  Sequel Ace
+//
+//  Created by James on 7/12/2020.
+//  Copyright © 2020 Sequel-Ace. All rights reserved.
+//
+
+import Foundation
+import os.log
+import Firebase
+
+
+/*
+reRequestSecureAccess
+addBookmark
+handle bookmarkDataIsStale
+revokeBookmark
+stopAllSecurityScopedAccess
+
+*/
+
+@objc final class SecureBookmarkManager: NSObject {
+
+	@objc static let sharedInstance = SecureBookmarkManager()
+
+    @objc public var bookmarks: [Dictionary<String, Data>] = []
+	@objc public var resolvedBookmarks: [URL] = []
+	@objc public var staleBookmarks: [String] = []
+
+    private let _NSURLBookmarkResolutionWithSecurityScope = URL.BookmarkResolutionOptions(rawValue: 1 << 10)
+    private let log: OSLog
+    private let prefs: UserDefaults = UserDefaults.standard
+    private var observer: NSKeyValueObservation?
+
+    private var iChangedTheBookmarks: Bool = false
+
+
+	override private init() {
+		log = OSLog(subsystem: Bundle.main.bundleIdentifier!, category: "secureBookmarks")
+
+        os_log("SecureBookmarkManager init.", log: log, type: .info)
+        Crashlytics.crashlytics().log("SecureBookmarkManager init.")
+
+		super.init()
+
+        // this manager *should* be the only thing changing the bookmarks pref, but in case...
+        observer = UserDefaults.standard.observe(\.SPSecureBookmarks, options: [.new,.old], changeHandler: { (defaults, change) in
+
+            if self.iChangedTheBookmarks == false{
+                os_log("SPSecureBookmarks changed NOT by SecureBookmarkManager.", log: self.log, type: .debug)
+                Crashlytics.crashlytics().log("SPSecureBookmarks changed NOT by SecureBookmarkManager.")
+                self.bookmarks.removeAll()
+                self.bookmarks = change.newValue!
+            }
+            // reset
+            self.iChangedTheBookmarks = false
+
+            // post notificay for ConnectionController
+            NotificationCenter.default.post(name: Notification.Name.init(NSNotification.Name.SPBookmarksChanged.rawValue), object: self)
+
+        })
+
+        // FIXME: @Kaspik need help ... need a guard or if let, don't want the default value...
+        bookmarks = prefs.array(forKey: SASecureBookmarks) as? [[String: Data]] ?? [["noBookmarks": Data()]]
```

Comment:
```
guard let secureBookmarks = prefs.array(forKey: SASecureBookmarks) as? [[String: Data]], secureBookmarks.isNotEmpty else {
	     os_log("Could not get secureBookmarks from prefs.", log: self.log, type: .error)
             Crashlytics.crashlytics().log("Could not get secureBookmarks from prefs.")
             return
}
```
will solve also fixme below

Btw, we should ideally migrate from array of dictionaries in preferences, to the NSData as we talked about? Not necessary here tho

Reply:
thanks! sorted!